File: gate.py
from relay import Relay
import time

class Gate:
    
    def __init__(self, relay_port):
        self.relay = Relay(relay_port)
        
    def open(self):
        self.relay.open()
        self.print_state()
        
    def close(self):
        self.relay.close()
        self.print_state()
        
    def toggle(self):
        self.relay.open()
        self.print_state()
        time.sleep(10)
        self.relay.close()
        self.print_state()
        
    def print_state(self):
        val = self.relay.get_value();
        if val == 0:
            print("{\"state\":\"open\"}")
        elif val == 1:
            print("{\"state\":\"close\"}")
        else:
            print("{\"error\":\"unable to determine state\"}")

File: gate.js
const {spawn} = require('child_process');

module.exports = class Gate {

  constructor(port) {
    this.port = port ? port : 26;
    this.onStateChangeListener = null;
  }

  setStateListener(fn) {
    this.onStateChangeListener = fn;
  }

  onStateChange(state) {
    if (this.onStateChangeListener) {
      this.onStateChangeListener(state);
    }
  }

  close() {
    spawn('python', [__dirname+'/gate_py_scripts/main.py', this.port, 'close']);
    this.getState();
  }

  open() {
    spawn('python', [__dirname+'/gate_py_scripts/main.py', this.port, 'open']);
    this.getState();
  }

  toggle() {
    spawn('python', [__dirname+'/gate_py_scripts/main.py', this.port, 'toggle']);
  }

  getState() {
    const cmd = spawn('python', [__dirname+'/gate_py_scripts/main.py', this.port, 'state']);
    cmd.stdout.on('data', (data) => {
      try {
        if (`${data}` !== '') {
          this.onStateChange(JSON.parse(`${data}`.trim()));
        }
      } catch(e) {
        console.log(e, data);
      }
      cmd.kill();
    });
  }

};